[Permissions API] Report the true permission state of Geolocation only when that API has been used since the page load

RupinMittal opened the pull request:
#### 386b21db7bdc2fc5135138027639ef8bf8a0402b
<pre>
[Permissions API] Report the true permission state of Geolocation only when that API has been used since the page load
<a href="https://bugs.webkit.org/show_bug.cgi?id=293089">https://bugs.webkit.org/show_bug.cgi?id=293089</a>
<a href="https://rdar.apple.com/151426960">rdar://151426960</a>

Reviewed by Youenn Fablet.

The current behavior of the Permissions API is:
- Return PROMPT in all cases where the permission state of the Geolocation API is DENIED.
- Return GRANTED in all cases where the permission state of the Geolocation API is GRANTED.

This PR changes WebKit&apos;s Permissions API to behave like so:
(On both MacOS and iOS).

User action                                   || Client tells WebKit || WebKit tells website    ||
----------------------------------------------||---------------------||-------------------------||
User says GRANTED to prompt once              || PROMPT              || GRANTED                 || geolocation requested by page since page load
                                              || PROMPT              || PROMPT                  || not requested since page load
User says GRANTED to prompt twice (24h grant) || PROMPT              || GRANTED                 || geolocation requested by page since page load
                                              || PROMPT              || PROMPT                  || not requested since page load
User says DENIED to prompt once               || PROMPT              || DENIED                  || geolocation requested by page since page load
                                              || PROMPT              || PROMPT                  || not requested since page load
User says DENIED to prompt twice (24h grant)  || PROMPT              || DENIED                  || geolocation requested by page since page load
                                              || PROMPT              || PROMPT                  || not requested since page load
User says GRANTED in permanent settings       || GRANTED             || GRANTED                 ||
User says DENIED in permanent settings        || DENIED              || DENIED                  || geolocation requested by page since page load
                                              || DENIED              || PROMPT (fingerprinting) || not requested since page load

The &quot;Client tells WebKit&quot; column represents how the client&apos;s implementation of the queryPermission
delegate behaves. Given this behavior, this PR makes changes so that the Permissions API behaves
like the &quot;WebKit tells website&quot; column.

WebKit&apos;s position is that the geolocation permission state should only be revealed to a site if
that site is willing to show a prompt asking for permission to use that API.

Prompts for geolocation are shown when one of the functions of the Geolocation API are called
(for example navigator.geolocation.getCurrentPosition). So the Permissions API should only reveal
the true permission state of the Geolocation API if such a function has been called since the page
load. If you reload the page, WebKit forgets that you&apos;ve tried to use the API.

The only exception to this rule is that if a user has permanently GRANTED--the user is indicating
that they trust this site, so the Permissions API will return the true state regardless of if the
site has attempted to get the geolocation. It&apos;s important to note that these settings (both permanent
and for a prompt) are per-site settings.

We don&apos;t make this exception in the case of permanently DENIED--as fingerprinting countermeasure,
we still return PROMPT in the case where Geolocation hasn&apos;t been requested.

Implementation details:

Navigator has a Geolocation object. On this object, we add a property (m_hasBeenRequested) which
gets set any time permission to get geolocation is requested. The Geolocation object is reset
every page load and so is this property.

When Permissions::query is called, it gets the permission state from the client. For geolocation,
clients return GRANTED or DENIED only in the cases where the user has granted or denied *permanently*.
If no permanent setting has been set, and user has simply granted or denied via a prompt, clients
return PROMPT. But if the user grants or denies via a prompt, the m_allowGeolocation property gets
set on the Geolocation object.

This means that Permissions::queryPermission can see that:
- If the result returned by client is GRANTED or DENIED --&gt; that&apos;s a permanent setting
- If the result returned by client is PROMPT and Geolocation has been requested --&gt; the user has
  responded to the prompt and their answer is contained in Geolocation::isAllowed()

The behavior in the table is tested by new API tests. These tests work on both MacOS and iOS.

* Source/WebCore/Modules/geolocation/Geolocation.cpp:
(WebCore::Geolocation::requestPermission):
(WebCore::Geolocation::resetIsAllowed):
* Source/WebCore/Modules/geolocation/Geolocation.h:

We add the new property m_hasBeenRequested here with a function to access it.

To access Geolocation&apos;s m_hasBeenRequested property, the Permissions API
code first gets the Geolocation from the NavigatorGeolocation. This is
supplement to the WebPage. Service workers don&apos;t have a WebPage and thus
don&apos;t have a NavigationGeolocation object. So we want to make sure
we don&apos;t try to createa Geolocation object in this case because attempting
to delete it later will result in a crash (we delete it by accessing it
from the supplement first--which doesn&apos;t exist in the worker case).

* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::query):
(WebCore::Permissions::determineGeolocationPermissionState):

This function decides if the permission state is a permanent one, or a temporary
user-response to a prompt. It contains the logic to change GRANTED and DENIED to
PROMPT in cases where the Geolocation API hasn&apos;t been used since the page load.
(The main logic of the above table is implemented here).

* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
(-[PermissionsAPIUIDelegate _webView:queryPermission:forOrigin:completionHandler:]):
(-[PermissionsAPIAndGeolocationRequestUIDelegate _webView:requestGeolocationPermissionForFrame:decisionHandler:]):
(TestWebKitAPI::TEST(PermissionsAPI, DataURL)):
(TestWebKitAPI::TEST(PermissionsAPI, OnChange)):
(TestWebKitAPI::testPermissionsAPIForGeolocation):
(TestWebKitAPI::TEST(PermissionsAPI, GeolocationPermissionGrantedFromPromptAndGeolcationRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolocationPermissionGrantedFromPromptAndGeolocationNotRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionDeniedFromPromptAndGeolocationRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionDeniedFromPromptAndGeolocationNotRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionGrantedPermanentlyAndGeolocationRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionGrantedPermanentlyAndGeolocationNotRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionDeniedPermanentlyAndGeolocationRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionDeniedPermanentlyAndGeolocationNotRequestedSinceLoad)):

Canonical link: <a href="https://commits.webkit.org/295237@main">https://commits.webkit.org/295237@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a034119ea8ba2ebf04ee4931cc9c372943196530

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104371 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14490 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109573 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55039 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106411 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24449 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32622 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79245 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107377 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19002 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94192 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59575 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18802 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12266 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54400 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88499 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12319 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111954 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31528 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23226 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88282 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31892 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90436 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87947 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22426 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32851 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10616 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/26995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31456 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36781 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31250 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34586 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32810 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -203,4 +217,31 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
+    RefPtr domWindow = document.domWindow();
+    if (!domWindow)
+        return std::nullopt;
+    RefPtr geolocation = NavigatorGeolocation::geolocation(domWindow->protectedNavigator());
```
geolocation can in theory be null.

RupinMittal replied:
Good catch! Interestingly it seems like `NavigatorGeolocation::geolocation()` will create a Geolocation object if one doesn't exist--and yet it returns a pointer. Since that pointer can't be null, I've changed the function to return a reference.

RupinMittal replied:
Seems like this change is causing a crash in `http/tests/permissions/permissions-query-shared-worker.html`. I'm changing it back and null-checking `geolocation` here.

cdumez replied:
I think the crashes are due to this code right here. This now constructs a `NavigatorGeolocation` and a `Geolocation` object for the summary Page of a ServiceWorker/SharedWorker. This is new since geolocation is not exposed to service / shared workers as far as I know.

I think we're crashing in `Geolocation::stopUpdating()` when calling `GeolocationController::from(page)` because the GeolocationController is set by the WebPage constructor and a service worker's dummy WebCore::Page likely doesn't have a corresponding WebPage object.

The fix likely has to do with not constructing NavigatorGeolocation / Geolocation for dummy service worker  shared worker pages.

cdumez replied:
This fixes the crashes for me by avoiding to construct the Geolocation object if it doesn't already exist:
```diff
diff --git a/Source/WebCore/Modules/geolocation/NavigatorGeolocation.cpp b/Source/WebCore/Modules/geolocation/NavigatorGeolocation.cpp
index 8b37216ff473..6a48fd70a2af 100644
--- a/Source/WebCore/Modules/geolocation/NavigatorGeolocation.cpp
+++ b/Source/WebCore/Modules/geolocation/NavigatorGeolocation.cpp
@@ -73,6 +73,14 @@ Geolocation* NavigatorGeolocation::geolocation(Navigator& navigator)
     return NavigatorGeolocation::from(navigator)->geolocation();
 }
 
+Geolocation* NavigatorGeolocation::optionalGeolocation(Navigator& navigator)
+{
+    NavigatorGeolocation* supplement = static_cast<NavigatorGeolocation*>(Supplement<Navigator>::from(&navigator, supplementName()));
+    if (!supplement)
+        return nullptr;
+    return supplement->m_geolocation.get();
+}
+
 Geolocation* NavigatorGeolocation::geolocation() const
 {
     if (!m_geolocation)
diff --git a/Source/WebCore/Modules/geolocation/NavigatorGeolocation.h b/Source/WebCore/Modules/geolocation/NavigatorGeolocation.h
index 7fd2074e1300..632fbdcc6ae7 100644
--- a/Source/WebCore/Modules/geolocation/NavigatorGeolocation.h
+++ b/Source/WebCore/Modules/geolocation/NavigatorGeolocation.h
@@ -39,6 +39,7 @@ public:
     static NavigatorGeolocation* from(Navigator&);
 
     static Geolocation* geolocation(Navigator&);
+    static Geolocation* optionalGeolocation(Navigator&);
     Geolocation* geolocation() const;
 
 #if PLATFORM(IOS_FAMILY)
diff --git a/Source/WebCore/Modules/permissions/Permissions.cpp b/Source/WebCore/Modules/permissions/Permissions.cpp
index b2c5f33d8c96..34dc50062fb7 100644
--- a/Source/WebCore/Modules/permissions/Permissions.cpp
+++ b/Source/WebCore/Modules/permissions/Permissions.cpp
@@ -241,19 +241,17 @@ static std::optional<PermissionState> determineGeolocationPermissionState(Permis
     if (!domWindow)
         return std::nullopt;
 
-    RefPtr geolocation = NavigatorGeolocation::geolocation(domWindow->protectedNavigator());
-    if (!geolocation)
-        return std::nullopt;
+    RefPtr geolocation = NavigatorGeolocation::optionalGeolocation(domWindow->protectedNavigator());
 
     switch (permissionState) {
     case PermissionState::Granted:
         return PermissionState::Granted;
     case PermissionState::Denied:
-        if (!geolocation->hasBeenRequested())
+        if (!geolocation || !geolocation->hasBeenRequested())
             return PermissionState::Prompt;
         return PermissionState::Denied;
     case PermissionState::Prompt:
-        if (!geolocation->hasBeenRequested())
+        if (!geolocation || !geolocation->hasBeenRequested())
             return PermissionState::Prompt;
         return geolocation->isAllowed() ? PermissionState::Granted : PermissionState::Denied;
     };
```

RupinMittal replied:
Sweet, thank you! I realized that this line is the problem but wasn't sure why exactly--your explanation definitely helps. I'll re-upload with the change.

youennf commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -660,6 +661,7 @@ void Geolocation::resetIsAllowed()
 {
     m_allowGeolocation = Unknown;
     revokeAuthorizationTokenIfNecessary();
+    m_hasBeenRequested = false;
```
Which case are you trying to cover?
If the page ever called geolocation, I think it is fine to keep exposing the permission state.

RupinMittal replied:
There isn't a specific case I'm trying to cover. I see a function called `Geolocation::resetAllGeolocationPermission()`-- I'm not wholly sure of what it's used for but it sounds like we want to reset our state, so I figured we should reset `m_hasBeenRequested` as well. What do you think?

youennf replied:
It seems this code path is executed when the page goes in page cache and is resumed.
That could help us decide what to do and maybe write a test for it.
It seems ok to keep this flag.

youennf commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -203,4 +217,31 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
+    case PermissionState::Prompt:
+        if (!geolocation->hasBeenRequested())
+            return PermissionState::Prompt;
+        return geolocation->isAllowed() ? PermissionState::Granted : PermissionState::Denied;
```
This is a difference between camera/microphone and geolocation.
If user denied access to camera/microphone through the prompt (non persistent) and getUserMedia is called again, we will deny getUserMedia automatically without a user gesture. With a user gesture, the user will be prompted again.
This looks good as is but maybe we should try to look at this inconsistency in the future.

RupinMittal replied:
Oh I see. This logic reflects the table. You mentioned that we can look at this in the future -- sounds good, perhaps we'll have to revisit the logic in the table if needed.

youennf replied:
Yes, let's look at this in a different bug report. Maybe we should file one to keep track of this.

youennf commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -203,4 +217,31 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
 
+#if ENABLE(GEOLOCATION)
+
+std::optional<PermissionState> Permissions::determineGeolocationPermissionState(PermissionState permissionState, const Document& document)
```
Can it be a free function? That would allow not taking protectedThis above?

RupinMittal replied:
It could and yes, that would allow us to not capture `protectedThis`. I was thinking about this too. I made it a separate function because it feels like `Permissions::query` is getting so big and this bit of logic encapsulates nicely into a function. It seems like a trade-off between making `Permissions::query` ever-larger and capturing `protectedThis`. I'm curious--how bad is it to capture `protectedThis`?

I can of course just stick the code into `Permissions::query` where the `determineGeolocationPermissionState` callsite is if we think that's better than capturing `protectedThis`. Let me know :)

RupinMittal replied:
Ah I just realized what you meant by free function--a lambda. That took me a minute. Yep, I have to use this `determineGeolocationPermissionState` in the worker case as well, and I can't use `protectedThis` there since we're posting task to another thread. So I'll make this a lambda and have that captured.

RupinMittal replied:
Oh `determineGeolocationPermissionState` uses the `document`. But in the worker case, we're not on the main thread--but `document` is a main thread object. So I can't use the document. What can I do here? @youennf @cdumez

youennf replied:
By free function, I just mean something like:
`static std::optional<PermissionState> determineGeolocationPermissionState(PermissionState permissionState, const Document& document) { ... }`

youennf replied:
As of worker vs. document, this is already handled since we are checking whether the context is a document or a worker (if a worker, we hop to main thread).
To better document this, we could add `ASSERT(isMainThread());` in line 157 and line 169.

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -153,4 +178,121 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
+TEST(PermissionsAPI, GeolcolationPermissionDeniedPermanentlyAndGeolocationNotRequestedSinceLoad)
+{
+    testPermissionsAPIForGeolocation(GeolocationPermissionState::DeniedFromPrompt, RequestGeolocationSincePageLoad::No, "prompt"_s);
+}
```
Tests look great!

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -153,4 +178,121 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
+
+TEST(PermissionsAPI, GeolcolationPermissionDeniedPermanentlyAndGeolocationRequestedSinceLoad)
+{
+    testPermissionsAPIForGeolocation(GeolocationPermissionState::DeniedFromPrompt, RequestGeolocationSincePageLoad::Yes, "denied"_s);
```
```suggestion
    testPermissionsAPIForGeolocation(GeolocationPermissionState::DeniedPermanently, RequestGeolocationSincePageLoad::Yes, "denied"_s);
```

RupinMittal replied:
Thank you for the catch! Fixed!

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -153,4 +178,121 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
+
+TEST(PermissionsAPI, GeolcolationPermissionDeniedPermanentlyAndGeolocationNotRequestedSinceLoad)
+{
+    testPermissionsAPIForGeolocation(GeolocationPermissionState::DeniedFromPrompt, RequestGeolocationSincePageLoad::No, "prompt"_s);
```
```suggestion
    testPermissionsAPIForGeolocation(GeolocationPermissionState::DeniedPermanently, RequestGeolocationSincePageLoad::No, "prompt"_s);
```

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -153,4 +178,121 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
+    EXPECT_STREQ(((NSString *)[scriptMessage body]).UTF8String, expectedResult.utf8().data());
+}
+
+TEST(PermissionsAPI, GeolocationPermissionGrantedFromPromptAndGeolcationRequestedSinceLoad)
```
```suggestion
TEST(PermissionsAPI, GeolocationPermissionGrantedFromPromptAndGeolocationRequestedSinceLoad)
```

youennf commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -153,4 +178,121 @@ static void urlEncodeIfNeeded(uint8_t byte, StringBuilder& buffer)
+    testPermissionsAPIForGeolocation(GeolocationPermissionState::GrantedFromPrompt, RequestGeolocationSincePageLoad::No, "prompt"_s);
+}
+
+TEST(PermissionsAPI, GeolcolationPermissionDeniedFromPromptAndGeolocationRequestedSinceLoad)
```
```suggestion
TEST(PermissionsAPI, GeolocolationPermissionDeniedFromPromptAndGeolocationRequestedSinceLoad)
```
Here and elsewhere.

RupinMittal replied:
Oops! Fixed!

youennf commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -203,4 +217,31 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
+    return std::nullopt;
+}
+
+#endif // ENABLE(GEOLCATION)
```
```suggestion
#endif // ENABLE(GEOLOCATION)
```

youennf commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -12074,8 +12074,6 @@ void WebPageProxy::queryPermission(const ClientOrigin& clientOrigin, const Permi
     } else if (descriptor.name == PermissionName::Geolocation) {
 #if ENABLE(GEOLOCATION)
         name = "geolocation"_s;
```
Maybe document that `shouldChangeDeniedToPrompt` is done in WebProcess directly?

youennf reviewed (approved):
LGTM

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
```diff
@@ -35,8 +38,18 @@
+static bool didReceiveGeolocationRequest;
+static bool requestGeolocationPermissionDelegateResult;
+
+enum class RequestGeolocationSincePageLoad { Yes, No };
```
Please, always `No` first so that No == 0 and Yes == 1.

cdumez replied:
Also `: bool`

RupinMittal commented on Source/WebCore/Modules/permissions/Permissions.cpp:
```diff
@@ -203,4 +233,34 @@ void Permissions::query(JSC::Strong<JSC::JSObject> permissionDescriptorValue, DO
+    if (!domWindow)
+        return std::nullopt;
+
+    RefPtr geolocation = NavigatorGeolocation::geolocation(domWindow->protectedNavigator());
```
This line is causing the crashes. If I don't create a Geolocation object here (and don't use it below of course), there's no crash. Not yet sure why.

cdumez commented on Source/WebCore/Modules/geolocation/Geolocation.cpp:
```diff
@@ -727,6 +729,11 @@ void Geolocation::stopUpdating()
+    auto* controller = GeolocationController::from(page);
+    if (!controller)
+        WTFLogAlways("GeolocationController is null");
+    RELEASE_ASSERT_WITH_MESSAGE(controller, "RELEASE_ASSERT - GeolocationController is null");
```
Don't forget to remove this debug logic from the PR as well.

RupinMittal committed:
[Permissions API] Report the true permission state of Geolocation only when that API has been used since the page load
https://bugs.webkit.org/show_bug.cgi?id=293089
rdar://151426960

Reviewed by Youenn Fablet.

The current behavior of the Permissions API is:
- Return PROMPT in all cases where the permission state of the Geolocation API is DENIED.
- Return GRANTED in all cases where the permission state of the Geolocation API is GRANTED.

This PR changes WebKit's Permissions API to behave like so:
(On both MacOS and iOS).

User action                                   || Client tells WebKit || WebKit tells website    ||
----------------------------------------------||---------------------||-------------------------||
User says GRANTED to prompt once              || PROMPT              || GRANTED                 || geolocation requested by page since page load
                                              || PROMPT              || PROMPT                  || not requested since page load
User says GRANTED to prompt twice (24h grant) || PROMPT              || GRANTED                 || geolocation requested by page since page load
                                              || PROMPT              || PROMPT                  || not requested since page load
User says DENIED to prompt once               || PROMPT              || DENIED                  || geolocation requested by page since page load
                                              || PROMPT              || PROMPT                  || not requested since page load
User says DENIED to prompt twice (24h grant)  || PROMPT              || DENIED                  || geolocation requested by page since page load
                                              || PROMPT              || PROMPT                  || not requested since page load
User says GRANTED in permanent settings       || GRANTED             || GRANTED                 ||
User says DENIED in permanent settings        || DENIED              || DENIED                  || geolocation requested by page since page load
                                              || DENIED              || PROMPT (fingerprinting) || not requested since page load

The "Client tells WebKit" column represents how the client's implementation of the queryPermission
delegate behaves. Given this behavior, this PR makes changes so that the Permissions API behaves
like the "WebKit tells website" column.

WebKit's position is that the geolocation permission state should only be revealed to a site if
that site is willing to show a prompt asking for permission to use that API.

Prompts for geolocation are shown when one of the functions of the Geolocation API are called
(for example navigator.geolocation.getCurrentPosition). So the Permissions API should only reveal
the true permission state of the Geolocation API if such a function has been called since the page
load. If you reload the page, WebKit forgets that you've tried to use the API.

The only exception to this rule is that if a user has permanently GRANTED--the user is indicating
that they trust this site, so the Permissions API will return the true state regardless of if the
site has attempted to get the geolocation. It's important to note that these settings (both permanent
and for a prompt) are per-site settings.

We don't make this exception in the case of permanently DENIED--as fingerprinting countermeasure,
we still return PROMPT in the case where Geolocation hasn't been requested.

Implementation details:

Navigator has a Geolocation object. On this object, we add a property (m_hasBeenRequested) which
gets set any time permission to get geolocation is requested. The Geolocation object is reset
every page load and so is this property.

When Permissions::query is called, it gets the permission state from the client. For geolocation,
clients return GRANTED or DENIED only in the cases where the user has granted or denied *permanently*.
If no permanent setting has been set, and user has simply granted or denied via a prompt, clients
return PROMPT. But if the user grants or denies via a prompt, the m_allowGeolocation property gets
set on the Geolocation object.

This means that Permissions::queryPermission can see that:
- If the result returned by client is GRANTED or DENIED --> that's a permanent setting
- If the result returned by client is PROMPT and Geolocation has been requested --> the user has
  responded to the prompt and their answer is contained in Geolocation::isAllowed()

The behavior in the table is tested by new API tests. These tests work on both MacOS and iOS.

* Source/WebCore/Modules/geolocation/Geolocation.cpp:
(WebCore::Geolocation::requestPermission):
(WebCore::Geolocation::resetIsAllowed):
* Source/WebCore/Modules/geolocation/Geolocation.h:

We add the new property m_hasBeenRequested here with a function to access it.

To access Geolocation's m_hasBeenRequested property, the Permissions API
code first gets the Geolocation from the NavigatorGeolocation. This is
supplement to the WebPage. Service workers don't have a WebPage and thus
don't have a NavigationGeolocation object. So we want to make sure
we don't try to createa Geolocation object in this case because attempting
to delete it later will result in a crash (we delete it by accessing it
from the supplement first--which doesn't exist in the worker case).

* Source/WebCore/Modules/permissions/Permissions.cpp:
(WebCore::Permissions::query):
(WebCore::Permissions::determineGeolocationPermissionState):

This function decides if the permission state is a permanent one, or a temporary
user-response to a prompt. It contains the logic to change GRANTED and DENIED to
PROMPT in cases where the Geolocation API hasn't been used since the page load.
(The main logic of the above table is implemented here).

* Source/WebCore/Modules/permissions/Permissions.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::queryPermission):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/PermissionsAPI.mm:
(-[PermissionsAPIUIDelegate _webView:queryPermission:forOrigin:completionHandler:]):
(-[PermissionsAPIAndGeolocationRequestUIDelegate _webView:requestGeolocationPermissionForFrame:decisionHandler:]):
(TestWebKitAPI::TEST(PermissionsAPI, DataURL)):
(TestWebKitAPI::TEST(PermissionsAPI, OnChange)):
(TestWebKitAPI::testPermissionsAPIForGeolocation):
(TestWebKitAPI::TEST(PermissionsAPI, GeolocationPermissionGrantedFromPromptAndGeolcationRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolocationPermissionGrantedFromPromptAndGeolocationNotRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionDeniedFromPromptAndGeolocationRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionDeniedFromPromptAndGeolocationNotRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionGrantedPermanentlyAndGeolocationRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionGrantedPermanentlyAndGeolocationNotRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionDeniedPermanentlyAndGeolocationRequestedSinceLoad)):
(TestWebKitAPI::TEST(PermissionsAPI, GeolcolationPermissionDeniedPermanentlyAndGeolocationNotRequestedSinceLoad)):

Canonical link: https://commits.webkit.org/295237@main
